# Problem Statement

#### It is important that credit card companies are able to recognize fraudulent credit card transactions so that customers are not charged for items that they did not purchase.

#### The datasets contains transactions made by credit cards in September 2013 by european cardholders. This dataset presents transactions that occurred in two days, where we have 492 frauds out of 284,807 transactions. The dataset is highly unbalanced, the positive class (frauds) account for 0.172% of all transactions. The objective to classify if a transaction is fraudlent.

# Import libraries

In [28]:
import pandas as pd
import seaborn as sns
import numpy as np

In [50]:
from scipy.stats import zscore
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix

In [29]:
credit = pd.read_csv('creditcard.csv')

In [30]:
credit.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
Time      284807 non-null float64
V1        284807 non-null float64
V2        284807 non-null float64
V3        284807 non-null float64
V4        284807 non-null float64
V5        284807 non-null float64
V6        284807 non-null float64
V7        284807 non-null float64
V8        284807 non-null float64
V9        284807 non-null float64
V10       284807 non-null float64
V11       284807 non-null float64
V12       284807 non-null float64
V13       284807 non-null float64
V14       284807 non-null float64
V15       284807 non-null float64
V16       284807 non-null float64
V17       284807 non-null float64
V18       284807 non-null float64
V19       284807 non-null float64
V20       284807 non-null float64
V21       284807 non-null float64
V22       284807 non-null float64
V23       284807 non-null float64
V24       284807 non-null float64
V25       284807 non-null float64
V26  

#### There are 31 columns in the dataset

In [32]:
credit.describe()

Time            V1            V2            V3            V4  \
count  284807.000000  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean    94813.859575  3.919560e-15  5.688174e-16 -8.769071e-15  2.782312e-15   
std     47488.145955  1.958696e+00  1.651309e+00  1.516255e+00  1.415869e+00   
min         0.000000 -5.640751e+01 -7.271573e+01 -4.832559e+01 -5.683171e+00   
25%     54201.500000 -9.203734e-01 -5.985499e-01 -8.903648e-01 -8.486401e-01   
50%     84692.000000  1.810880e-02  6.548556e-02  1.798463e-01 -1.984653e-02   
75%    139320.500000  1.315642e+00  8.037239e-01  1.027196e+00  7.433413e-01   
max    172792.000000  2.454930e+00  2.205773e+01  9.382558e+00  1.687534e+01   

                 V5            V6            V7            V8            V9  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean  -1.552563e-15  2.010663e-15 -1.694249e-15 -1.927028e-16 -3.137024e-15   
std    1.380247e+00  1.332271e+00  1.237094e+00  1.194353e+00  1.098632e+00   
min   -1.137433e+02 -2.616051e+01 -4.355724e+01 -7.321672e+01 -1.343407e+01   
25%   -6.915971e-01 -7.682956e-01 -5.540759e-01 -2.086297e-01 -6.430976e-01   
50%   -5.433583e-02 -2.741871e-01  4.010308e-02  2.235804e-02 -5.142873e-02   
75%    6.119264e-01  3.985649e-01  5.704361e-01  3.273459e-01  5.971390e-01   
max    3.480167e+01  7.330163e+01  1.205895e+02  2.000721e+01  1.559499e+01   

       ...           V21           V22           V23           V24  \
count  ...  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   ...  1.537294e-16  7.959909e-16  5.367590e-16  4.458112e-15   
std    ...  7.345240e-01  7.257016e-01  6.244603e-01  6.056471e-01   
min    ... -3.483038e+01 -1.093314e+01 -4.480774e+01 -2.836627e+00   
25%    ... -2.283949e-01 -5.423504e-01 -1.618463e-01 -3.545861e-01   
50%    ... -2.945017e-02  6.781943e-03 -1.119293e-02  4.097606e-02   
75%    ...  1.863772e-01  5.285536e-01  1.476421e-01  4.395266e-01   
max    ...  2.720284e+01  1.050309e+01  2.252841e+01  4.584549e+00   

                V25           V26           V27           V28         Amount  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  284807.000000   
mean   1.453003e-15  1.699104e-15 -3.660161e-16 -1.206049e-16      88.349619   
std    5.212781e-01  4.822270e-01  4.036325e-01  3.300833e-01     250.120109   
min   -1.029540e+01 -2.604551e+00 -2.256568e+01 -1.543008e+01       0.000000   
25%   -3.171451e-01 -3.269839e-01 -7.083953e-02 -5.295979e-02       5.600000   
50%    1.659350e-02 -5.213911e-02  1.342146e-03  1.124383e-02      22.000000   
75%    3.507156e-01  2.409522e-01  9.104512e-02  7.827995e-02      77.165000   
max    7.519589e+00  3.517346e+00  3.161220e+01  3.384781e+01   25691.160000   

               Class  
count  284807.000000  
mean        0.001727  
std         0.041527  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 31 columns]

In [33]:
credit.drop(columns='Time', inplace=True)

In [36]:
y=credit.pop('Class')

# Normalise the data

In [37]:
X=credit.apply(zscore)

In [38]:
X.head()

V1        V2        V3        V4        V5        V6        V7  \
0 -0.694242 -0.044075  1.672773  0.973366 -0.245117  0.347068  0.193679   
1  0.608496  0.161176  0.109797  0.316523  0.043483 -0.061820 -0.063700   
2 -0.693500 -0.811578  1.169468  0.268231 -0.364572  1.351454  0.639776   
3 -0.493325 -0.112169  1.182516 -0.609727 -0.007469  0.936150  0.192071   
4 -0.591330  0.531541  1.021412  0.284655 -0.295015  0.071999  0.479302   

         V8        V9       V10  ...       V20       V21       V22       V23  \
0  0.082637  0.331128  0.083386  ...  0.326118 -0.024923  0.382854 -0.176911   
1  0.071253 -0.232494 -0.153350  ... -0.089611 -0.307377 -0.880077  0.162201   
2  0.207373 -1.378675  0.190700  ...  0.680975  0.337632  1.063358  1.456320   
3  0.316018 -1.262503 -0.050468  ... -0.269855 -0.147443  0.007267 -0.304777   
4 -0.226510  0.744326  0.691625  ...  0.529939 -0.012839  1.100011 -0.220123   

        V24       V25       V26       V27       V28    Amount  
0  0.110507  0.246585 -0.392170  0.330892 -0.063781  0.244964  
1 -0.561131  0.320694  0.261069 -0.022256  0.044608 -0.342475  
2 -1.138092 -0.628537 -0.288447 -0.137137 -0.181021  1.160686  
3 -1.941027  1.241904 -0.460217  0.155396  0.186189  0.140534  
4  0.233250 -0.395202  1.041611  0.543620  0.651816 -0.073403  

[5 rows x 29 columns]

# Split the data into training and test set

In [39]:
X_train, X_test, y_train, y_test=train_test_split(X, y, test_size=0.3)

In [41]:
X_train, X_val, y_train, y_val=train_test_split(X_train, y_train, test_size=0.3)

In [42]:
len(X_train)

139554

In [43]:
len(X_test)

85443

In [44]:
len(X_val)

59810

In [40]:
model_tree=DecisionTreeClassifier()

In [45]:
model_tree.fit(X_train, y_train)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=None, splitter='best')

In [46]:
model_tree.score(X_train,y_train)

1.0

In [47]:
model_tree.score(X_val,y_val)

0.9991640193947501

In [49]:
y_predict=model_tree.predict(X_test)

In [51]:
confusion_matrix(y_test, y_predict)

array([[85268,    44],
       [   27,   104]], dtype=int64)

In [53]:
len(y_test[y_test==0])

85312

In [54]:
len(y_test[y_test==1])

131

In [55]:
len(y_predict[y_predict==0])

85295

In [56]:
len(y_predict[y_predict==1])

148

In [57]:
(104/131) * 100

79.38931297709924

### From the total 131 fradulent transactions in the test set the model accuratley predicted 104 transactions as fradulent which is 79% correct. This is not a very good score. A good model should predict the fradulent transactions more accurately even if it makes a mistake by classifying non-fradulent transactions as fraudulent.

In [124]:
model_tree1=DecisionTreeClassifier(class_weight='balanced', max_leaf_nodes=25)

In [125]:
model_tree1.fit(X_train, y_train)

DecisionTreeClassifier(class_weight='balanced', criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=25,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=None, splitter='best')

In [126]:
y_predict=model_tree1.predict(X_test)

In [127]:
confusion_matrix(y_test, y_predict)

array([[84080,  1232],
       [   15,   116]], dtype=int64)

In [119]:
(116/131)*100

88.54961832061069

### With some hyper parameter tuning, now from the total 131 fradulent transactions in the test set the model accuratley predicted 116 transactions as fradulent which is 88.5% correct. This is a better score.